src/manifolds/SymplecticStiefel.jl

```diff
@@ -0,0 +1,664 @@
+    V_inv = lu!(add_scaled_I!(symplectic_inverse_times(M, q, p), 1))
+
+    X .= 2 .* ((p / V_inv .- p / U_inv) .+ ((p + q) / U_inv) .- p)
+    return X
+end
+
+@doc raw"""
+    manifold_dimension(::Symplectic{n})
+
+Returns the dimension of the symplectic Stiefel manifold embedded in ``ℝ^{2n \times 2k}``,
+i.e. [^Bendokat2021]
+````math
+    \operatorname{dim}(\operatorname{SpSt}(2n, 2k)) = (4n - 2k + 1)k.
+````
+"""
+manifold_dimension(::SymplecticStiefel{n,k}) where {n,k} = (4n - 2k + 1) * k
+
+@doc raw"""
+    project(::Union{SymplecticStiefel,Symplectic}, p, A)
+    project!(::Union{SymplecticStiefel, Symplectic}, Y, p, A)
+
+Given a point ``p \in \operatorname{SpSt}(2n, 2k)``,
+project an element ``A \in \mathbb{R}^{2n \times 2k}`` onto
+the tangent space ``T_p\operatorname{SpSt}(2n, 2k)`` relative to
+the euclidean metric of the embedding ``\mathbb{R}^{2n \times 2k}``.
+
+That is, we find the element ``X \in T_p\operatorname{SpSt}(2n, 2k)``
+which solves the constrained optimization problem
+
+````math
+    \operatorname{min}_{X \in \mathbb{R}^{2n \times 2k}} \frac{1}{2}||X - A||^2, \quad
+    \text{s.t.}\;
+    h(X) = X^T Q p + p^T Q X = 0,
+````
+where ``h : \mathbb{R}^{2n \times 2k} \rightarrow \operatorname{skew}(2k)`` defines
+the restriction of ``X`` onto the tangent space ``T_p\operatorname{SpSt}(2n, 2k)``.
+"""
+project(::Union{SymplecticStiefel,Symplectic}, p, A)
+
+function project!(::Union{SymplecticStiefel,Symplectic}, Y, p, A)
+    Q = SymplecticMatrix(Y, p, A)
+    Q_p = Q * p
+
+    function h(X)
+        XT_Q_p = X' * Q_p
+        return XT_Q_p .- XT_Q_p'
+    end
+
+    # Solve for Λ (Lagrange mutliplier):
+    pT_p = p' * p  # (2k × 2k)
+    Λ = sylvester(pT_p, pT_p, h(A) ./ 2)
+
+    Y[:, :] = A .- Q_p * (Λ .- Λ')
+    return Y
+end
+
+@doc raw"""
+    rand(M::SymplecticStiefel{n, k})
+
+Generate a random point ``p \in \operatorname{SpSt}(2n, 2k)``
+by first generating a random symplectic matrix
+``p_{\operatorname{Sp}} \in \operatorname{Sp}(2n)``,
+and then projecting onto the Symplectic Stiefel manifold using the
+[`canonical_projection`](@ref).
+That is, ``p = π(p_{\operatorname{Sp}})``.
+"""
+function Base.rand(M::SymplecticStiefel{n,k}; hamiltonian_norm=1 / 2) where {n,k}
+    p_symplectic = rand(Symplectic(2n); hamiltonian_norm=hamiltonian_norm)
+    return canonical_projection(M, p_symplectic)
+end
+
+@doc raw"""
+    random_vector(::SymplecticStiefel{n, k}, p)
+
+The symplectic Stiefel tangent space at ``p`` can be parametrized as [^Bendokat2021]
+````math
+    T_p\operatorname{SpSt}(2n, 2k) = \{X = pΩ + p^sB \;|\;
+        Ω ∈ ℝ^{2k × 2k}, Ω^+ = -Ω,
+        p^s ∈ \operatorname{SpSt}(2n, 2(n- k)), B ∈ ℝ^{2(n-k) × 2k}, \},
+````
+where ``Ω \in \mathfrak{sp}(2n,F)`` is Hamiltonian and ``p^s`` means
+the symplectic complement of ``p`` s.t. ``p^{+}p^{s} = 0``.
+
+To then generate random tangent vectors at ``p``, we set ``B = 0`` and generate a random
+Hamiltonian matrix ``Ω``.
+"""
+function random_vector(
```

Comment:
I've talked with @kellertuer on Slack about random vector generation and we agreed to be a bit more conservative about our interface. So the preferred way to get a random vector is `rand(M::AbstractManifold; vector_at=nothing)`, that is getting a random tangent vector at `p` will be `rand(M; vector_at=p)`. This avoids conflict with the API of `Base.rand` that leaves one argument for the distribution, but we can pass in keyword arguments whatever we like. `random_vector` as written here should stay but be internal (not exported) and called by `rand(M::AbstractManifold)` when `vector_at` is not `nothing`, more or less like this:
```julia
function Base.rand(M::SymplecticStiefel; vector_at=nothing, hamiltonian_norm = (vector_at === nothing ? 1/2 : 1.0))
    if vector_at === nothing
        # code for generation of random points
    else
        random_vector(M, vector_at; hamiltonian_norm = hamiltonian_norm)
    end
end
```

Reply:
Very nice, I also talked with @kellertuer about this today and it now sounds like you've found a good interface for this 
sort of functionality :relaxed:

Reply:
I've made the changes to this API-style now.